miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/RealMiniAppDisplay.kt

```
@@ -1,97 +1,57 @@
 
-    override fun getMiniAppView(): View = this
+    override suspend fun getMiniAppView(activityContext: Context): View? =
+        provideMiniAppWebView(activityContext)
 
     @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
     override fun destroyView() {
-        stopLoading()
-        webViewClient = null
-        destroy()
+        miniAppWebView?.destroyView()
+        miniAppWebView = null
     }
 
-    override fun runSuccessCallback(callbackId: String, value: String) {
-        post {
-            evaluateJavascript(
-                "MiniAppBridge.execSuccessCallback(\"$callbackId\", \"$value\")"
-            ) {}
-        }
-    }
-
-    override fun runErrorCallback(callbackId: String, errorMessage: String) {
-        post {
-            evaluateJavascript(
-                "MiniAppBridge.execErrorCallback(\"$callbackId\", \"$errorMessage\")"
-            ) {}
-        }
-    }
-
-    private fun getWebViewAssetLoader() = WebViewAssetLoader.Builder()
-        .setDomain(miniAppDomain)
-        .addPathHandler(
-            "/$SUB_DOMAIN_PATH/", WebViewAssetLoader.InternalStoragePathHandler(
-                context,
-                File(basePath)
-            )
-        )
-        .addPathHandler(
-            "/", WebViewAssetLoader.InternalStoragePathHandler(
-                context,
-                File(basePath)
-            )
-        )
-        .build()
+    private suspend fun provideMiniAppWebView(activityContext: Context): View? =
+        if (isContextValid(activityContext)) {
+            if (miniAppWebView != null)
+                miniAppWebView
+            else {
+                withContext(Dispatchers.Main) {
+                    MiniAppWebView(
+                        context = activityContext,
+                        basePath = basePath,
+                        appId = appId,
+                        miniAppMessageBridge = miniAppMessageBridge
+                    )
+                }
+            }
+        } else null
```

If they provided the wrong context, then this is a case where they've implemented the SDK wrong. So maybe this should throw an error instead of returning null? It would allow `getMiniAppView` to have a non-null return type, and it would give them a helpful error message to see what they've done wrong. What do you think?

---

I think we should keep it nullable here for `destroyView` purpose 🤔